pkg/cmd/pulumi/destroy.go

```diff
@@ -257,3 +286,52 @@ func newDestroyCmd() *cobra.Command {
 
 	return cmd
 }
+
+// seperateProtected returns a list or unprotected and protected resources
+// respectively. This allows us to safely destroy all resources in the
+// unprotected list without invalidating any resource in the protected list.
+// Protection is contravarient.
+//
+// A
+// B: Parent = A
+// C: Parent = A, Protect = True
+//
+// -->
+//
+// Unprotected: B
+// Protected: A, C
+//
+// We rely on the fact that `resources` is topologically sorted with respect to
+// its dependencies. This function understands that providers live outside this
+// topological sort.
+func seperateProtected(resources []*resource.State) (
+	/*unprotected*/ []*resource.State /*protected*/, []*resource.State) {
+	dg := graph.NewDependencyGraph(resources)
+	transitiveProtected := graph.ResourceSet{}
+	for _, r := range resources {
+		if r.Protect {
+			rProtected := dg.TransitiveDependenciesOf(r)
+			rProtected[r] = true
+			transitiveProtected.UnionWith(rProtected)
+		}
+	}
+	allResources := graph.NewResourceSetFromArray(resources)
+	return allResources.SetMinus(transitiveProtected).ToArray(), transitiveProtected.ToArray()
+}
+
+// Returns the number of protected resources that remain. Appends all unprotected resources to `targetUrns`.
+func handleExcludeProtected(s backend.Stack, targetUrns *[]resource.URN) (int, error) {
```

Comment:
Nit (this is mostly fine, perhaps I have bad FP habits here): This one signature is a bit funky (C-style?) it needs targetUrns to be empty when started, and it populates it when done, so why not  move `*[]resource.URN` to the result of the function rather than a parameter?